components/script/dom/vrdisplay.rs

```diff
@@ -0,0 +1,618 @@
+        api_sender.send(CanvasMsg::WebVR(msg)).unwrap();
+    }
+}
+
+impl VRDisplay {
+    fn webvr_thread(&self) -> Option<IpcSender<WebVRMsg>> {
+        self.global().as_window().webvr_thread()
+    }
+
+    pub fn get_display_id(&self) -> u64 {
+        self.display.borrow().0.display_id
+    }
+
+    pub fn update_display(&self, display: &webvr::VRDisplayData) {
+        self.display.borrow_mut().0 = display.clone();
+        if let Some(ref stage) = display.stage_parameters {
+            if self.stage_params.get().is_none() {
+                let params = Some(VRStageParameters::new(&stage, &self.global()));
+                self.stage_params.set(params.as_ref().map(|v| v.deref()));
+            } else {
+                self.stage_params.get().unwrap().update(&stage);
+            }
+        } else {
+            self.stage_params.set(None);
+        }
+
+    }
+
+    pub fn handle_webvr_event(&self, event: &webvr::VRDisplayEvent) {
+        match *event {
+            webvr::VRDisplayEvent::Connect(ref display) => {
+                self.update_display(&display);
+            },
+            webvr::VRDisplayEvent::Disconnect(_id) => {
+                self.display.borrow_mut().0.connected = false;
+            },
+            webvr::VRDisplayEvent::Activate(ref display, _) |
+            webvr::VRDisplayEvent::Deactivate(ref display, _) |
+            webvr::VRDisplayEvent::Blur(ref display) |
+            webvr::VRDisplayEvent::Focus(ref display) => {
+                self.update_display(&display);
+                self.notify_event(&event);
+            },
+            webvr::VRDisplayEvent::PresentChange(ref display, presenting) => {
+                self.update_display(&display);
+                self.presenting.set(presenting);
+                self.notify_event(&event);
+            },
+            webvr::VRDisplayEvent::Change(ref display) => {
+                // Change event doesn't exist in WebVR spec.
+                // So we update display data but don't notify JS.
+                self.update_display(&display);
+            }
+        };
+    }
+
+    fn notify_event(&self, event: &webvr::VRDisplayEvent) {
+        let root = Root::from_ref(&*self);
+        let event = VRDisplayEvent::new_from_webvr(&self.global(), &root, &event);
+        event.upcast::<Event>().fire(self.upcast());
+    }
+
+    fn init_present(&self) {
+        self.presenting.set(true);
+        let (sync_sender, sync_receiver) = ipc::channel().unwrap();
+        *self.frame_data_receiver.borrow_mut() = Some(sync_receiver);
+
+        let display_id = self.display.borrow().0.display_id;
+        let api_sender = self.layer_ctx.get().unwrap().ipc_renderer();
+        let js_sender = self.global().script_chan();
+        let address = Trusted::new(&*self);
+        let near_init = self.depth_near.get();
+        let far_init = self.depth_far.get();
+
+        thread::Builder::new().name("WebVR_RAF".into()).spawn(move || {
```

Comment:
It's must be one per display, because it waits VSync for the specific display. I'll improve the documentation ;)